src/pip/_internal/cache.py

```diff
@@ -168,11 +172,17 @@ def get(
         # type: (...) -> Link
         candidates = []
 
+        canonical_package_name = None
+        if package_name:
+            canonical_package_name = canonicalize_name(package_name)
         for wheel_name in self._get_candidates(link, package_name):
             try:
                 wheel = Wheel(wheel_name)
             except InvalidWheelFilename:
                 continue
+            assert canonical_package_name
```

Comment:
This seems strange with the `package_name,    # type: Optional[str]` possibly equal to `None`.

Reply:
@xavfernandez it's because `_get_candidates` returns an empty if `package_name` is `None` (`can_not_cache`). So when it enters the loop `package_name` must be set.

Reply:
It might be worth explaining this in a comment :)

Or maybe simply put:
```
if not package_name:
    return link
```
at the start of the function (I agree that this could happen in an other PR)

Reply:
@xavfernandez an early exit is simpler indeed. Done in a separate commit, as well as another small simplification. Thanks for the review.